File: Deepspace_ruby/Deepspace/lib/SpaceStation.rb
#encoding:utf-8

require_relative 'SpaceStationToUI'

module Deepspace

class SpaceStation
  @@MAXFUEL=100
  @@SHIELDLOSSPERUNITSHOT=0.1

  attr_reader :ammoPower, :fuelUnits, :name, :nMedals, :shieldPower, :hangar

  def initialize(n,s)
    @name = n
    @ammoPower = s.ammoPower
    @fuelUnits = s.fuelUnits
    @shieldPower = s.shieldPower
    @hangar = nil
    @weapons = []
    @shieldBoosters = []
    @pendingDamage = nil
    @nMedals = nil
  end

  def getUIversion
    SpaceStationToUI.new(self)
  end

  def receiveWeapon(w)
    if @hangar != nil
      @hangar.addWeapon(w)
    end
  end

  def receiveShieldBooster(s)
    if @hangar != nil
        @hangar.addShieldBooster(s)
    end
  end

  def receiveHangar(h)
    if @hangar == nil
      @hangar = h
    end
  end

  def discardHangar
    @hangar = nil
  end

  def receiveSupplies(s)
    @ammoPower += s.ammoPower
    @shieldPower += s.shieldPower
    if @fuelUnits + s.fuelUnits <= @@MAXFUEL
        @fuelUnits += s.fuelUnits
    else
        @fuelUnits = @@MAXFUEL
    end
  end

  def setPendingDamage(d)
    @pendingDamage = d.adjust(@weapons, @shieldBoosters)
  end

  def mountWeapon(i)
    if @hangar != nil
      rm = @hangar.removeWeapon(i)
      if rm != nil
        @weapons.push(rm)
      end
    end
  end

  def mountShieldBooster(i)
    if @hangar != nil
      rm = @hangar.removeShieldBooster(i)
      if rm != nil
        @shieldBoosters.push(rm)
      end
    end

  end

  def discardWeaponInHangar(i)
    if @hangar != nil
      @hangar.removeWeapon(i)
    end
  end

  def discardShieldBoosterInHangar(i)
    if @hangar != nil
      @hangar.removeShieldBooster(i)
    end

  end

  def getSpeed
    return @fuelUnits.to_f / @@MAXFUEL
  end

  def move
    @fuelUnits -= getSpeed * @fuelUnits.to_f
  end

  def validState
    return @pendingDamage == nil || @pendingDamage.hasNoEffect
  end

  def cleanUpMountedItems
    #Guardamos los índices
    indices = []
    for i in 0...@weapons.length
      if @weapons[i].uses == 0
        indices.push(i)
      end
    end

    indices.sort.reverse

    for i in indices
      @weapons.delete_at(i)
    end

    indices = []

    for i in 0...@shieldBoosters.length
      if @shieldBoosters[i].uses == 0
        indices.push(i)
      end
    end

    indices.sort.reverse

    for i in indices
      @shieldBoosters.delete_at(i)
    end
  end

  #Para la siguiente práctica
    def fire
      size = @weapons.length
      factor = 1.to_f
      for i in 0...size
        factor *= @weapons[i].useIt
      end

      return @ammoPower*factor
    end

    def protection
      size = @shieldBoosters.length
      factor = 1.to_f
      for i in 0...size
        factor *= @shieldBoosters[i].useIt
      end
      return @shieldPower*factor
    end

    def receiveShot(shot)
      myProtection = @currentStation.protection
      if(myProtection >= shot)
        @shieldPower -= @@SHIELDLOSSPERUNITSHOOT * shot
        @shieldPower = [0.0,@shieldPower].max
        return ShotResult::RESIST
      else
        @shieldPower = 0.0
        return ShotResult::DONOTRESIST
      end
    end

    def setLoot(loot)
      dealer = CardDealer.new
      h = loot.nHangars
      if h > 0
        hangar = dealer.nextHangar
        receiveHangar(hangar)
      end

      elements = loot.nSupplies
      for i in 1..elements
        sup = dealer.nextSuppliesPackage
        receiveSupplies(sup)
      end

      elements = loot.nWeapons
      for i in 1..elements
        weap = dealer.nextWeapon
        receiveWeapon(weap)
      end

      elements = loot.nShields
      for i in 1..elements
        sh = dealer.nextShieldBooster
        receiveShieldBooster(sh)
      end

      medals = loot.nMedals
      @nMedals += medals
    end


    def discardWeapon(i)
      size = @weapons.length
      if i >= 0 && i < size
        w = @weapons.delete_at(i)
        if @pendingDamage != nil
          @pendingDamage.discardWeapon(w)
          cleanPendingDamage
        end
      end
    end

    def discardShieldBooster(i)
       size = @shieldBoosters.length
      if i >= 0 && i < size
        s = @shieldBoosters.delete_at(i)
        if @pendingDamage != nil
          @pendingDamage.discardShieldBooster
          cleanPendingDamage
        end
      end

    end

  def to_s
     return "Name #{@name} AmmoPower: #{@ammoPower} FuelUnits: #{@fuelUnits}
     ShieldPower: #{@shieldPower} Medals: #{@nMedals} Weapons: #{@weapons}
     ShieldBoosters: #{@shieldBoosters} Hangar: #{@hangar} PendingDamage: #{@pendingDamage}"
   end

  private
    def assignFuelValue(f)
      if f <= @@MAXFUEL
        @fuelsUnits = f
      end
    end

    def cleanPendingDamage
      if @pendingDamage.hasNoEffect
        @pendingDamage = nil
      end
    end


end
end

File: Deepspace_java/Deepspace/src/deepspace/SpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
import java.util.function.Predicate;
import java.lang.Math.*;

/**
 *
 * @author victor
 */
public class SpaceStation {
    private static int MAXFUEL = 100; 
    private static float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private float ammoPower, fuelUnits, shieldPower;
    private int nMedals;
    private String name;
    private Damage pendingDamage;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;
    
    
    SpaceStation(String n, SuppliesPackage s) {
        name = n;
        ammoPower = s.getAmmoPower();
        fuelUnits = s.getFuelUnits();
        shieldPower = s.getShieldPower();
    }
    private void assignFuelValue(float f) {
      if(f<=  MAXFUEL)
        fuelUnits = f;
    } 
    
    
    private void cleanPendingDamage(){
        if(pendingDamage.hasNoEffect())
             pendingDamage = null;
    }
    
    SpaceStationToUI getUIversion() {
        return new SpaceStationToUI(this);
    }

    public float getAmmoPower() {
        return ammoPower;
    }

    public float getFuelUnits() {
        return fuelUnits;
    }

    public float getShieldPower() {
        return shieldPower;
    }

    public int getNMedals() {
        return nMedals;
    }

    public String getName() {
        return name;
    }

    public Damage getPendingDamage() {
        return pendingDamage;
    }

    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }

    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }

    public Hangar getHangar() {
        return hangar;
    }
       
    public void discardHangar() {
      hangar = null;             
    }
    
    public boolean receiveShieldBooster(ShieldBooster s) {
        if(hangar != null)
            return hangar.addShieldBooster(s);
               
        return false;
    }
    
    public boolean receiveWeapon(Weapon w) {
        if(hangar != null)
            return hangar.addWeapon(w);
               
        return false;
    }
    
    public void receiveHangar(Hangar h) {
        if(hangar != null)
            hangar = h;
    }
    
    public void receiveSupplies(SuppliesPackage s) {
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        if(fuelUnits + s.getFuelUnits() <= MAXFUEL)
            fuelUnits += s.getFuelUnits();
        else
             fuelUnits = MAXFUEL;
    }
    
    public void setPendingDamage(Damage d) {
        pendingDamage = d;
    }
    
    public void mountWeapon(int i) {
        if(hangar != null)
            weapons.add(hangar.removeWeapon(i));
    }
    
    public void mountShieldBooster(int i) {
        if(hangar != null)
            shieldBoosters.add(hangar.removeShieldBooster(i));
    }
    
    public void discardWeaponInHangar(int i) {
        if(hangar != null)
           hangar.removeWeapon(i);
    }
    
    public void discardShieldBoosterInHangar(int i) {
        if(hangar != null)
           hangar.removeShieldBooster(i);
    }
    
    public float getSpeed() {
        return fuelUnits / MAXFUEL;        
    }
    
    public void move() {
        fuelUnits -= getSpeed() * fuelUnits;
    }
    
    public boolean validState() {
        return pendingDamage == null || pendingDamage.hasNoEffect();
    }
    
    public void cleanUpMountedItems() {
      //ESTO PUEDE QUE NO FUNCIONE ASÍ
      Predicate<Weapon> weaponPredicate;
        weaponPredicate = (w-> w.getUses() == 0);
      weapons.removeIf(weaponPredicate);
      
      Predicate<ShieldBooster> shieldPredicate;
        shieldPredicate = (s-> s.getUses() == 0);
      shieldBoosters.removeIf(shieldPredicate);      
    }
    
    public float fire() { 
        int size = weapons.size();
        float factor = 1;
        for (int i = 0; i < size; i++) {
            factor *= weapons.get(i).useIt();
        }        
        return factor;  
    }
    
    public float protection() { 
        int size = shieldBoosters.size();
        float factor = 1;
        for (int i = 0; i < size; i++) {
            factor *= shieldBoosters.get(i).useIt();
        }        
        return factor;
    }
    
    public ShotResult receiveShot(float shot) { 
        float myProtection = protection();
        if(myProtection >= shot) {
            shieldPower -= SHIELDLOSSPERUNITSHOT * shot;
            shieldPower = (float) Math.max(0.0,shieldPower); //Hay que hacer un cast, sino da error de conversion ya que max devuelve double
            return ShotResult.RESIST;
        } else {
            shieldPower = (float) 0.0;
            return ShotResult.DONOTRESIST;
        }   
    }
    public void setLoot(Loot loot) { 
        CardDealer dealer = CardDealer.getInstance();
        int h = loot.getNHangars();
        if(h > 0) {
            Hangar hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }
        
        int elements = loot.getNSupplies();
        for(int i = 0; i < elements; i++) {
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        for(int i = 0; i < elements; i++) {
            Weapon weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }
        
        elements = loot.getNShields();
        for(int i = 0; i < elements; i++) {
            ShieldBooster sh = dealer.nextShieldBooster();
            receiveShieldBooster(sh);
        }
        
        int medals = getNMedals();
        nMedals += medals;    
    }
    
    public void discardWeapon(int i) { 
        int size = weapons.size();
        if(i >= 0 && i < size) {
            Weapon w = weapons.remove(i);
            if(pendingDamage != null) {
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
        }
    }    
    
    
    public void discardShieldBooster(int i) { 
        int size = shieldBoosters.size();
        if(i >= 0 && i < size) {
            ShieldBooster s = shieldBoosters.remove(i);
            if(pendingDamage != null) {
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();                
            }
        }    
    }   
       
    
}
